ZOOKEEPER-2819:Changing membership configuration via rolling restart …

hanm opened the pull request:
…does not work on 3.5.x.

This patch disables the creation of dynamic config files (zoo.cfg.dynamic) and static config back up files (zoo.cfg.bak) when the dynamic reconfig feature flag (reconfigEnabled) is disabled. With this patch the membership information (such as server list) will be stored in static zoo.cfg file and such information would not go through quorum and leader / follower sync phase, which makes it possible for users to continue using the old rolling restart approach.

@shralex PTAL.

shralex commented on src/java/main/org/apache/zookeeper/server/quorum/QuorumPeer.java:
```diff
@@ -1744,9 +1744,14 @@ public void setAcceptedEpoch(long e) throws IOException {
-       QuorumVerifier prevQV = setQuorumVerifier(qv, true);
+       // and restart leader election if config changed. If reconfig feature is
+       // not enabled, just set last committed quorum verifier and bail out.
+       QuorumVerifier prevQV = setQuorumVerifier(qv, isReconfigEnabled);
```
The server acts according to the config it has in its quorum verifier, so setting quorum verifier overwrites the config file settings with the leader's settings

hanm replied:
Updated. Now I am returning immediately at the start of processReconfig if the feature flag is not set.

asdf2014 commented on src/java/test/org/apache/zookeeper/server/quorum/ReconfigRollingRestartCompatibilityTest.java:
```diff
@@ -0,0 +1,94 @@
+            mt[i].shutdown();
+        }
+    }
+}
```
Should add a new line for the end of file.

hanm replied:
Should be fixed now.

hanm commented:
Thanks for review, @shralex @asdf2014 . Code updated. 

I also refactored a couple of reconfig tests to consolidate the setting of the feature flag, which makes the test both easier to read and also fix potential race conditions I found out while working on this patch: because the flag was set in a different thread while the peers are running, it is possible the peer thread still cache old value. Such case would not happen for real use case (aside from artificial tests we created here) because the feature flag will not change after ensemble is started.

shralex commented:
Michael, thanks for the changes. Do we need some tests that can catch the issue I pointed out - one server is getting rebooted with a new config file, but other servers push old config to that server, need to make sure that it actually acts upon or has the config its supposed to have. Two scenarios - config pushed durning leader sync, and config pushed during leader election (in FastLeaderElection.java). Perhaps you're already doing this, I haven't looked.

hanm commented:
Added test cases to cover rolling restart. The pre-commit Jenkins failure is a known flaky test.

shralex commented on src/java/test/org/apache/zookeeper/server/quorum/ReconfigRecoveryTest.java:
```diff
@@ -28,9 +28,15 @@
 public class ReconfigRecoveryTest extends QuorumPeerTestBase {
+    @Before
+    public void setup() {
+        QuorumPeerConfig.setReconfigEnabled(true);
```
is the same needed also in all other reconfig test files ?

hanm replied:
Yes if those tests require reconfig feature being enabled. I think I've patched all reconfig test in ZOOKEEPER-2014, and the modifications of related tests here are all about refactoring and consolidating setting of the flag in a single place.

shralex commented on src/java/test/org/apache/zookeeper/test/ReconfigExceptionTest.java:
```diff
@@ -89,6 +89,7 @@ public void tearDown() throws Exception {
 
     @Test(timeout = 10000)
     public void testReconfigDisabledByDefault() throws InterruptedException {
+        QuorumPeerConfig.setReconfigEnabled(false);
```
Given that the test is called "disabled-by-default" Its strange to disable this manually here. I suggest to rename the test to reflect what its doing - checking that when the flag is false reconfig ops aren't allowed.

shralex commented on src/java/test/org/apache/zookeeper/server/quorum/ReconfigRollingRestartCompatibilityTest.java:
```diff
@@ -0,0 +1,275 @@
+    }
+
+    // A successful write operation indicates a quorum is up.
+    private void verifyQuorum(int sid)
```
Can you reuse some helper methods from ReconfigTest.java ?

hanm replied:
I missed testServerHasConfig - I think that should do it.

shralex commented on src/java/test/org/apache/zookeeper/server/quorum/ReconfigRollingRestartCompatibilityTest.java:
```diff
@@ -0,0 +1,275 @@
+        for (int i = 0; i < serverCount; ++i) {
+            mt[i].shutdown();
+            mt[i].start();
+            verifyQuorum(i);
```
two comments:
1. is calling start immediately after shutdown effective ? I don't remember what shutdown is doing, but if its setting a flag and a thread periodically checks it, it may take time to take effect.
2. nothing is changing about the config as far as I can see. Why would this ever fail ?

hanm replied:
for 1 - for this specific case this is fine, because this test case does not change setting of the reconfig flag (from false to true or vice versa) so there is no concern on that end. shutdown is synchronous so once shutdown finishes, it's safe to call start. 

I think your concern was around the visibility of a shared variable among multiple threads (in this case, reconfigEnabled). Adding volatile should address the concern (also a new test case specifically to test the visibility of the variable is required.). I will do this as part of sub-task ZOOKEEPER-2828.

For 2, yes as the test name implies, no config changes here. I guess I can remove the config verify part here though leave it there for insurance seems ok too.

shralex commented on src/java/test/org/apache/zookeeper/server/quorum/ReconfigRollingRestartCompatibilityTest.java:
```diff
@@ -0,0 +1,275 @@
+        Set<String> expectedConfigs = new HashSet<>();
+        for (String config : serverAddress.values()) {
+            // Remove "server.x=" prefix.
+            expectedConfigs.add(config.substring(9));
```
this seems fragile (works only for single digit ids, perhaps use the index of "="

hanm replied:
With the ReconfigTest helpers I can remove these code, hopefully.

shralex commented on src/java/test/org/apache/zookeeper/server/quorum/ReconfigRollingRestartCompatibilityTest.java:
```diff
@@ -0,0 +1,275 @@
+        Set<String> expectedConfigs = new HashSet<>();
+        for (String conf : oldServerAddress.values()) {
+            // Remove "server.x=" prefix.
+            expectedConfigs.add(conf.substring(9));
```
same comment - using 9 seems fragile. Also perhaps create a method for this since this code appears twice

shralex commented on src/java/test/org/apache/zookeeper/server/quorum/ReconfigRollingRestartCompatibilityTest.java:
```diff
@@ -0,0 +1,275 @@
+        serverCount = serverAddress.size();
+        Assert.assertEquals("Server count should be 5 after config update.", serverCount, 5);
+
+        mt = Arrays.copyOf(mt, mt.length + 2);
```
Please add a comment here to say that the new servers should have a config with all five servers, and the old servers a config with 3 servers.

hanm commented:
>> One question - what happens with getConfig when reconfig is disabled ? Does it still work ?

Yes getConfigwill continue working until the heat death of the universe regardless of reconfigEnabled flag settings :).

hanm commented:
Addressed code review comments from @shralex:

* Use ReconfigTest.testServerHasConfig for additional test coverage.
* Improve fragile substring extraction.
* Add comments around test logic.
* Also made a fix to ensure during rolling restart, config information returned from getConfig is consistent with local config.

hanm commented on src/java/main/org/apache/zookeeper/server/quorum/Learner.java:
```diff
@@ -374,6 +374,13 @@ else if (qp.getType() == Leader.SNAP) {
+                // ZOOKEEPER-2819: overwrite config node content extracted
+                // from leader snapshot with local config, to avoid potential
+                // inconsistency of config node content during rolling restart.
+                if (!QuorumPeerConfig.isReconfigEnabled()) {
```
@shralex instead of avoiding deserializing the config zNode we could overwrite the deserialized zNode content with local quorum config. Test will be covered in the test improvement JIRA ZOOKEEPER-2828.

shralex commented on src/java/test/org/apache/zookeeper/server/quorum/ReconfigRollingRestartCompatibilityTest.java:
```diff
@@ -0,0 +1,263 @@
+        for (int i = 0; i < serverCount; ++i) {
+            mt[i].shutdown();
+            mt[i].start();
+            verifyQuorumConfig(i, joiningServers, leavingServers);
```
you can remove the leavingServers variable and pass null here. Same in the other test

shralex commented on src/java/test/org/apache/zookeeper/server/quorum/ReconfigRollingRestartCompatibilityTest.java:
```diff
@@ -0,0 +1,263 @@
+
+        Map<Integer, String> oldServerAddress = new HashMap<>(serverAddress);
+        List<String> newServers = new ArrayList<>(joiningServers);
+        newServers.add("3"); newServers.add("4");
```
move this one line down until after you have the actual addresses, then put the actual config lines of 3 and 4 into newServers

shralex commented on src/java/test/org/apache/zookeeper/server/quorum/ReconfigRollingRestartCompatibilityTest.java:
```diff
@@ -0,0 +1,263 @@
+import java.util.HashSet;
+import java.util.ArrayList;
+
+import org.apache.zookeeper.ZooKeeper;
```
you don't really need leavingServers - you can pass null instead of it in all the tests since no one is leaving

hanm commented:
@shralex The snapshot and sync tests will be covered in ZOOKEEPER-2828. I will update the PR shortly to address other comments.

hanm commented:
Thanks @shralex for review. Good catch on the "3", "4" server issue. PR updated that addresses latest comments. The failed test is a known flaky one.

shralex commented:
LGTM
